src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs

```diff
@@ -230,8 +243,38 @@ public void RunClassInitialize(TestContext testContext)
                 throw new NullReferenceException(Resource.TestContextIsNull);
             }
 
-            // If class initialization is not done, then do it.
-            if (!this.IsClassInitializeExecuted)
+            MethodInfo initializeMethod = null;
+            string failedClassInitializeMethodName = string.Empty;
+
+            try
+            {
+                // If class initialization is not done, we can proceed to perform
+                // base class initialization first
+                if (!this.IsClassInitializeExecuted)
+                {
+                    // ClassInitialize methods for base classes are called in reverse order of discovery
+                    // Base -> Child TestClass
+                    var baseClassInitializeStack = new Stack<Tuple<MethodInfo, MethodInfo>>(this.BaseClassInitAndCleanupMethods);
+                    while (baseClassInitializeStack.Count > 0)
+                    {
+                        initializeMethod = baseClassInitializeStack.Pop().Item1;
+                        if (initializeMethod != null)
+                        {
+                            initializeMethod?.InvokeAsSynchronousTask(null, testContext);
+                            this.ExecutedBaseClassInitializeMethods.Add(initializeMethod);
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                this.ClassInitializationException = ex;
+                failedClassInitializeMethodName = initializeMethod.Name;
+            }
+
+            // If class initialization is not done and class initialize method is not null,
+            // and class initialization exception is null, then do it.
+            if (!this.IsClassInitializeExecuted && this.classInitializeMethod != null && this.ClassInitializationException == null)
             {
                 // Aquiring a lock is usually a costly operation which does not need to be
```

Comment:
Does the base class init need the same locking mechanism that the current class init uses below? Would be useful when tests are run in parallel.

Reply:
You're right. Didn't considered it in the first place when changing the RunClassInitialize method for base class init.
So, taking the same locking logic for the base class init methods, considering parallel runs. thanks for that catch.